framework/src/relationshipmanagers/AugmentSparsityOnInterface.C

```diff
@@ -78,79 +91,132 @@ AugmentSparsityOnInterface::operator()(const MeshBase::const_element_iterator &
-  // solve
-  if (_use_displaced_mesh)
+  // If we're on a dynamic mesh or we have not yet constructed the mortar mesh, we need to ghost the
+  // entire interface because we don't know a priori what elements will project onto what. We *do
+  // not* add the whole interface if we are a coupling functor because it is very expensive. This is
+  // because when building the sparsity pattern, we call through to the ghosting functors with one
+  // element at a time (and then below we do a loop over all the mesh's active elements). It's
+  // perhaps faster in this case to deal with mallocs coming out of MatSetValues, especially if the
+  // mesh displacements are relatively small
+  if ((!_amg || _use_displaced_mesh) && !_is_coupling_functor)
   {
-    for (const auto & elem : _mesh.getMesh().active_element_ptr_range())
+    for (const Elem * const elem : _mesh->active_element_ptr_range())
     {
-      if (elem->subdomain_id() == _subdomain_pair.first ||
-          elem->subdomain_id() == _subdomain_pair.second)
+      if (generating_mesh)
       {
-        if (elem->processor_id() != p)
+        // We are still generating the mesh, so it's possible we don't even have the right boundary
+        // ids created yet! So we actually ghost all boundary elements and all lower dimensional
+        // elements who have parents on a boundary
+        if (elem->on_boundary())
           coupled_elements.insert(std::make_pair(elem, null_mat));
-        auto ip = elem->interior_parent();
-        if (ip->processor_id() != p)
-          coupled_elements.insert(std::make_pair(ip, null_mat));
+        else if (const Elem * const ip = elem->interior_parent())
+        {
+          if (ip->on_boundary())
+            coupled_elements.insert(std::make_pair(elem, null_mat));
+        }
+      }
+      else
+      {
+        // We've finished generating our mesh so we can be selective and only ghost elements lying
+        // in our lower-dimensional subdomains and their interior parents
+
+        mooseAssert(primary_boundary_id != Moose::INVALID_BOUNDARY_ID,
+                    "Primary boundary id should exist by now. If you're using a MeshModifier "
+                    "please use the corresponding MeshGenerator instead");
+        mooseAssert(secondary_boundary_id != Moose::INVALID_BOUNDARY_ID,
+                    "Secondary boundary id should exist by now. If you're using a MeshModifier "
+                    "please use the corresponding MeshGenerator instead");
+        mooseAssert(primary_subdomain_id != Moose::INVALID_BLOCK_ID,
+                    "Primary subdomain id should exist by now. If you're using a MeshModifier "
+                    "please use the corresponding MeshGenerator instead");
+        mooseAssert(secondary_subdomain_id != Moose::INVALID_BLOCK_ID,
+                    "Secondary subdomain id should exist by now. If you're using a MeshModifier "
+                    "please use the corresponding MeshGenerator instead");
+
```

Comment:
Roy styles, Assert everywhere , haha

Reply:
It helps me keep track of what I'm doing and I grow more and more paranoid over the years :smile: